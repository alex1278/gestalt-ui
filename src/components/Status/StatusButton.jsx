import React from 'react';
import PropTypes from 'prop-types';
import MenuButton from 'components/Menus/MenuButton';
import MoreVertIcon from '@material-ui/icons/MoreVert';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import statusMap from './statusMap';

const StatusButton = ({ children, status, asButton, ...rest }) => {
  const backgroundColor = asButton ? statusMap(status).color : null;

  const icon = asButton
    ? <ExpandMoreIcon fontSize="small" color="action" />
    : <MoreVertIcon fontSize="small" color="primary" />;

  return (
    <MenuButton
      {...rest}
      id="container-actions-menu"
      icon={icon}
      iconAfter
      flat={asButton}
      flatVariant="outlined"
      flatColor={backgroundColor}
      label={status}
      disabled={!status}
    >
      {children}
    </MenuButton>
  );
};

StatusButton.propTypes = {
  children: PropTypes.any,
  status: PropTypes.string,
  asButton: PropTypes.bool,
};

StatusButton.defaultProps = {
  children: null,
  status: 'PENDING',
  asButton: false,
};

export default StatusButton;
