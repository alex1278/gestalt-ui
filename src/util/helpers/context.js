/**
 * generateContextEntityState
 * @param {*} params
 */
export function generateContextEntityState(params) {
  const entity = {
    id: params.environmentId || params.workspaceId || null,
    key: null,
  };

  if (params.workspaceId && params.environmentId) {
    entity.key = 'environments';
  }

  if (params.workspaceId && !params.environmentId) {
    entity.key = 'workspaces';
  }

  // if (params.workspaceId && params.environmentId && params.providerId) {
  //   entity.id = params.providerId;
  //   entity.key = 'providers';
  // }

  return entity;
}
