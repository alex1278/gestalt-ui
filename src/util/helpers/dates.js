import moment from 'moment';

export const getNow = () => moment();

export const getAgo = (num = 0, unit = 'day') =>
  moment().subtract(num, unit);

export const isRecentWithin = (timestamp, num, unit) =>
  moment(timestamp).isBetween(getAgo(num, unit), getNow());
