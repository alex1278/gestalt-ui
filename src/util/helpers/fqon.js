// sad hack for parsing url

export default function fqonParent(organization) {
  if (organization.org) {
    const urlParser = document.createElement('a');
    urlParser.href = organization.org.href;
    return urlParser.pathname.replace('/', '').split('/')[0];
  }

  return organization;
}
