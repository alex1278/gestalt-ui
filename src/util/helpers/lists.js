export function insertItem(array, payload) {
  return [...array, payload];
}

export function removeItem(array, payload) {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a === payload), 1);

  return newArray;
}

export function removeItemById(array, id, keyField = 'id') {
  const newArray = array.slice();
  newArray.splice(newArray.findIndex(a => a[keyField] === id), 1);

  return newArray;
}

export function updateItem(array, payload) {
  return array.map((item) => {
    if (item.id === payload.id) {
      return {
        ...payload,
      };
    }

    // Leave every other item unchanged
    return item;
  });
}
