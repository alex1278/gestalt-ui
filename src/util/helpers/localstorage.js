/**
 * getItem
 * @param {*} key
 */
export const getItem = (key) => {
  try {
    return localStorage.getItem(key);
  } catch (err) {
    return undefined;
  }
};

/**
 * saveItem
 * @param {*} key
 * @param {*} value
 */
export const saveItem = (key, value) => {
  try {
    localStorage.setItem(key, value);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};

/**
 * removeItem
 * @param {*} key
 * @param {*} value
 */
export const removeItem = (key) => {
  try {
    localStorage.removeItem(key);
  } catch (err) {
    // ignore write errors
  }

  return undefined;
};
