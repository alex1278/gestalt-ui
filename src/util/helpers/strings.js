import { startCase, toLower } from 'lodash';

/**
 * getLastFromSplit
 * Parses and returns the last item in a structure like GrandPareng::Parent::Child
 * @param {String} string
 */
export function getLastFromSplit(string, delimiter = '::') {
  if (string) {
    const split = string.split(delimiter);

    return split[split.length - 1];
  }
  return string;
}

/**
 * truncate
 * @param {String} string
 * @param {Number} at
 * @param {String} trailing
 */
export function truncate(string, at, trailing = '...') {
  if (string) {
    if (string.length < at) {
      return string;
    }

    return `${string.substring(0, at).trim()}${trailing}`;
  }

  return string;
}

/**
 * toTitleCase
 * @param {String} string
 */
export function toTitleCase(string) {
  return startCase(toLower(string));
}


// TODO: Will be rmeoved when we have security config
export const checkIfPassword = field =>
  field && field.length && (field.toUpperCase().includes('PASSWORD') || field.toUpperCase().includes('SECRET') || field.toUpperCase().includes('KEY'));

export default {
  getLastFromSplit,
  truncate,
  toTitleCase,
  checkIfPassword
};

export const removeHostFromURL = (url) => {
  if (url) {
    return url.replace(/^.*\/\/[^/]+:?[0-9]?\//i, '');
  }

  return url;
};

export const isBase64 = (str) => {
  try {
    return btoa(atob(str)) === str;
  } catch (err) {
    return false;
  }
};
