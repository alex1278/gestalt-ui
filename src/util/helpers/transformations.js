import map from 'lodash/map';
import compact from 'lodash/compact';
import get from 'lodash/get';

/**
 * arrayToMap
 * @param {*} array
 * @param {*} keyName
 * @param {*} valueName
 */
export function arrayToMap(array = [], keyName = 'name', valueName = 'value') {
  if (!Array.isArray(array)) {
    return array;
  }

  return Object.assign({}, ...array.map(v => ({ [v[keyName]]: v[valueName] || '' })));
}

/**
 * mapTo2DArray
 * @param {*} object
 * @param {*} keyName
 * @param {*} valueName
 * @param {*} mergeSet
 */
export function mapTo2DArray(object = {}, keyName = 'name', valueName = 'value', mergeSet = {}) {
  if (typeof object !== 'object') {
    return object;
  }

  return map(object, (value, key) => (Object.assign({ [keyName]: key, [valueName]: value || '' }, mergeSet)));
}

/**
 * stringDemiltedToArray
 * convert comma delimited string to an array and remove blank entries
 * @param {*} string
 */
export function stringDemiltedToArray(string) {
  return compact(string.split(','));
}

/**
 * convertFromMaps
 * used to convert env vars from a map to an array and reconcile inherited vs own
 * @param {*} own
 * @param {*} inherited
 */
export function convertFromMaps(own = {}, inherited = {}, keyName = 'name', valueName = 'value') {
  const ownVars = mapTo2DArray(own, keyName, valueName, { inherited: false });

  // base case just return the mapped own vars
  if (!Object.keys(inherited)) {
    return ownVars;
  }

  const inheritedVars = mapTo2DArray(inherited, keyName, valueName, { inherited: true, overridden: false });

  const reconcileInheritedVars = inheritedVars.map((i) => {
    // find the dupe value in ownVars
    const index = ownVars.findIndex(o => o[keyName] === i[keyName]);

    // if the value is differnt mark the matching var as overriden
    if (index > -1 && ownVars[index][valueName] !== i[valueName]) {
      return { ...i, [valueName]: ownVars[index][valueName], inherited: true, overridden: true };
    }

    // default case the own env var
    return i;
  });

  const results = [...reconcileInheritedVars];

  // "merge" the ownVars back in
  ownVars.forEach((v) => {
    const index = inheritedVars.findIndex(o => o[keyName] === v[keyName]);

    if (index === -1) {
      results.push(v);
    }
  });

  return results;
}

/**
 * Builds an object from a delimited string
 * @param {String} string
 * @param {String} delimiter
 */
export function nestedObjectFromString(string, value = null, delimeter = '.') {
  const [last, ...paths] = string.split(delimeter).reverse();
  return paths.reduce((acc, el) => ({ [el]: acc }), { [last]: value });
}


export const getRoutePattern = (object, urlPattern) => {
  const keys = [];

  urlPattern.split('/').forEach((item) => {
    const search = item.startsWith(':');

    if (search) {
      keys.push(get(object, item.slice(1)));
    } else {
      keys.push(item);
    }
  });

  return keys.join('/');
};
