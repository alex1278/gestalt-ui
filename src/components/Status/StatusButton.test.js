import React from 'react';
import StatusButton from './StatusButton';
import { renderWithTheme } from '../../../test/helpers';

describe('Components::StatusButton::StatusButton', () => {
  it('mounts with default props', () => {
    const { container } = renderWithTheme(<StatusButton><div>hi</div></StatusButton>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when a status is passed', () => {
    const { container } = renderWithTheme(<StatusButton status="running"><div>hi</div></StatusButton>);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly when the button has inMenu passed', () => {
    const { container } = renderWithTheme(<StatusButton status="running" inMenu><div>hi</div></StatusButton>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
