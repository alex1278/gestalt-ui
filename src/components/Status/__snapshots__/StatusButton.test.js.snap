// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::StatusButton::StatusButton mounts with default props 1`] = `
<button
  aria-haspopup="true"
  aria-label="More"
  class="MuiButtonBase-root-8 MuiIconButton-root-2 IconButtonStyled-button-1"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label-7"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root-11 MuiSvgIcon-colorPrimary-12 MuiSvgIcon-fontSizeSmall-18"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root-24"
  />
</button>
`;

exports[`Components::StatusButton::StatusButton renders correctly when a status is passed 1`] = `
<button
  aria-haspopup="true"
  aria-label="More"
  class="MuiButtonBase-root-8 MuiIconButton-root-2 IconButtonStyled-button-1"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label-7"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root-11 MuiSvgIcon-colorPrimary-12 MuiSvgIcon-fontSizeSmall-18"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root-24"
  />
</button>
`;

exports[`Components::StatusButton::StatusButton renders correctly when the button has inMenu passed 1`] = `
<button
  aria-haspopup="true"
  aria-label="More"
  class="MuiButtonBase-root-8 MuiIconButton-root-2 IconButtonStyled-button-1"
  tabindex="0"
  type="button"
>
  <span
    class="MuiIconButton-label-7"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root-11 MuiSvgIcon-colorPrimary-12 MuiSvgIcon-fontSizeSmall-18"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
  </span>
  <span
    class="MuiTouchRipple-root-24"
  />
</button>
`;
